pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing CRD")
+	c.queue.Add(key)
+}
+
+func (c *controller) enqueueFromAPIBinding(obj interface{}) {
+	binding, ok := obj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), binding)
+
+	for _, boundResource := range binding.Status.BoundResources {
+		key := kcpcache.ToClusterAwareKey(apibinding.ShadowWorkspaceName.String(), "", boundResource.Schema.UID)
+		logging.WithQueueKey(logger, key).V(2).Info("queueing CRD via APIBinding")
+		c.queue.Add(key)
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := c.getCRD(clusterName, name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	logger := logging.WithObject(klog.FromContext(ctx), obj)
+	ctx = klog.NewContext(ctx, logger)
+
+	if !apihelpers.IsCRDConditionTrue(obj, apiextensionsv1.Established) {
+		logger.V(4).Info("CRD is not established", "conditions", fmt.Sprintf("%#v", obj.Status.Conditions))
+		return nil
+	}
+
+	result, err := c.getAPIBindingsByBoundResourceUID(obj.Name)
+	if err != nil {
+		return err
+	}
+
+	// Index is used to quickly return
+	if len(result) > 0 {
+		return nil
+	}
+
+	// The index can't solely be relied on to be certain this CRD can be deleted.
```

Comment:
I don't think this is doing what you think this is doing. Right now the `c.listAPIBindings()` call is using the lister - because of the way the machinery works, whether you access the list of bindings from the lister & filter manually using Go code, or if you are accessing the filtered list via the indexer, you are fundamentally operating against the exact same set of objects in the in-memory cache, and you will never (under normal circumstances) get a different result. I think you might want to use the client to do a live call here, instead? It may be useful then also to add labels or something to the bound CRDs for efficient server-side selection, as CRDs are large objects and having to do a full LIST here will be a heavy call.

Reply:
My comment here is misleading.  I'm not listing here due to a concern about resources in in the index being out of date due to an unprocessed event, etc.  I'm instead seeing if any `APIBindings` are in an unready state to avoid deleting the CRD immediately after it was created.  The `APIBinding` controller will wait for the bound CRD to become established before it adds the corresponding `BoundResource` to its status, so it's expected that the case where an `APIBinding` has no `BoundResources` will be common.  [This thread](https://github.com/kcp-dev/kcp/pull/2298#discussion_r1014475991) is discussing this as well. 

Note that if _any_ `APIBinding` is not ready, we cannot confidently delete any CRD.  

If that all sounds reasonable to you, then I'll update the comment in the code to be less confusing.

Reply:
So I started writing the index, and then I realized that there was another edge case: the `APIExport` reference in an `APIBinding` is invalid.  This makes the index much less clean than "by phase", some thing like "APIExportIsValidAndPhaseIsNotBound".  So I kept the list in place but added a big comment trying to explain why we are doing the list after the index lookup.

Reply:
I'm not sure it's clear why this new edge case is related to the index.  

> Note that if any APIBinding is not ready, we cannot confidently delete any (bound) CRD.

The edge case make this statement untrue since `APIBindings` with an invalid `APIExport` will never become ready, and we must exclude them from consideration to prevent every bound CRD from being requeued forever.